### DISCUSSION TITLE
[Proposal] Apply copy-on-write (COW) rule wherever possible

### DISCUSSION THREAD

**Discussion by pwwang** [pwwang/datar#103] [General]
Only copy the data when possible. For example:

```python
df = mutate(mtcars)

# currently, df is a copy of mtcars
# but if there are no changes made to mtcars, it should be:
# df is mtcars
```

**Comment by machow** [comment]
Hey--I suspect that copying ends up being pretty cheap computationally and space-wise. AFAIK a lot of the internal references are preserved (e.g. to underlying "Blocks" / arrays), until there's something that would mutate them.

(I could definitely be wrong here though)

**Comment by pwwang** [comment]
Thanks for pointing this out. I somehow agree with you. This is a low priority as this could anyway save resources.
